### PR TITLE
add user settings

### PR DESCRIPTION
**Pull request by andxu**

### CODE REVIEW

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaDebugDelegateCommandHandler.java [hunk 1]
```diff
@@ -32,6 +33,8 @@
 
     public static String CONFIG_LOG_LEVEL = "vscode.java.configLogLevel";
 
+    public static String UPDATE_USER_SETTINGS = "vscode.java.updateUserSettings";
```
Comment (akaroml) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JavaDebugDelegateCommandHandler.java]: We need to not confuse the user by providing the command vscode.java.updateUserSetting. It's actually debugger settings. So suggest making it "vscode.java.updateDebuggerSettings".

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/formatter/StringObjectFormatter.java [hunk 2]
```diff
@@ -26,7 +26,7 @@
 
 public class StringObjectFormatter extends ObjectFormatter implements IValueFormatter {
     public static final String MAX_STRING_LENGTH_OPTION = "max_string_length";
-    private static final int DEFAULT_MAX_STRING_LENGTH = 100;
+    private static final int DEFAULT_MAX_STRING_LENGTH = 200;
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/formatter/StringObjectFormatter.java]: Can I understand why 100 -> 200?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/EvaluateRequestHandler.java [hunk 3]
```diff
@@ -61,18 +59,11 @@ public void handle(Command command, Arguments arguments, Response response, IDeb
...
+        final boolean showStaticVariables = UserSettings.showStaticVariables;
+
         Map<String, Object> options = context.getVariableFormatter().getDefaultOptions();
-        if (evalArguments.format != null && evalArguments.format.hex) {
-            options.put(NumericFormatter.NUMERIC_FORMAT_OPTION, NumericFormatEnum.HEX);
-        }
-        if (showFullyQualifiedNames) {
-            options.put(SimpleTypeFormatter.QUALIFIED_CLASS_NAME_OPTION, showFullyQualifiedNames);
-        }
+        VariableUtils.applyFormatterOptions(options, evalArguments.format != null && evalArguments.format.hex,
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/EvaluateRequestHandler.java]: Do we need to apply the formatter options every time a evaluate request is processed? If the options stay inside VadiableUtils, then we can call it when a user actually changes the settings.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/VariablesRequestHandler.java [hunk 4]
```diff
@@ -57,18 +55,11 @@ public void handle(Command command, Arguments arguments, Response response, IDeb
...
-        boolean showStaticVariables = true;
-        // TODO: When vscode protocol support customize settings of value format, showFullyQualifiedNames should be one of the options.
-        boolean showFullyQualifiedNames = true;
-        if (varArgs.format != null && varArgs.format.hex) {
-            options.put(NumericFormatter.NUMERIC_FORMAT_OPTION, NumericFormatEnum.HEX);
-        }
-        if (showFullyQualifiedNames) {
-            options.put(SimpleTypeFormatter.QUALIFIED_CLASS_NAME_OPTION, showFullyQualifiedNames);
-        }
+        VariableUtils.applyFormatterOptions(options, varArgs.format != null && varArgs.format.hex, false);
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/VariablesRequestHandler.java]: Same here, do we really need to apply settings during processing a request?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/formatter/StringObjectFormatter.java (hunk 2 above)
Comment (andxu) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/formatter/StringObjectFormatter.java]: the default length is the length most users saw, the 100 is too small for normal cases like connections string, path, I would advise to use 256.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/EvaluateRequestHandler.java (hunk 3 above)
Comment (andxu) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/EvaluateRequestHandler.java]: The condition for generating options is also influenced by argument, not only user settings.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugSettings.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                    String jsonStr = (String) arg;
+
+                    try {
+                        Map<String, Object> map = JsonUtils.fromJson(jsonStr, new HashMap<String, Object>().getClass());
+
+                        for (String keyStr : map.keySet()) {
+                            Object valueObj = map.get(keyStr);
+                            try {
+                                switch (keyStr) {
+                                    case "show_hex":
```
Comment (testforstephen) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/DebugSettings.java]: Not consistent with the key on vscode user settings.

**Pull request merged**